### DISCUSSION TITLE
Parallizing using Joblib

### DISCUSSION THREAD

**Discussion by cyasing** [pybamm-team/PyBaMM#2893] [Q&A]
I set up a loop for multiple cells being given different current inputs using Joblib, like so:
```
for i, t in enumerate(t_eval):
    if t == 0:
        model = [pybamm.lithium_ion.SPM(options)] * N_cells
    else:
        for i in range(N_cells):
            old_model_i[i] = model[i]
            model[i] = old_model_i[i].set_initial_conditions_from(solution[i], inplace=True)

    solution = Parallel(n_jobs=-2)(delayed(cell_simulation)(t, i, n, param, model, I_batt) for n in range(N)) # Joblib code
```
where the cell_simulation function is:
```
def cell_simulation(t, i, n, param, model, I_batt):
    solver = pybamm.CasadiSolver(mode="fast")
    sim = pybamm.Simulation(model=model[n], parameter_values=param[i], solver=solver)
    solution = sim.solve([t,t+60],inputs={"Current function [A]" : I_batt[n][i]})
    del solver
    return solution
```
I was able to get perfect Joblib run not even a couple months ago, and now I retry and it hits me with:
`pybamm.expression_tree.exceptions.ModelError: model is underdetermined (too many variables)`.
If I run the program normally without parallelizing, it works perfectly. 

Did something change in pybamm recently to have such an error or should I look into Joblib?

**Comment by TomTranter** [comment]
Not sure about joblib. However, liionpack has all the parallel architecture in place so an option could be instead of running a connected pack we run an ensemble. Probably wouldn't take long to code up

**Reply by cyasing** [reply]
Does liionpack allow using casadi/dfn solver? Last I remember it solved only using ElectrodeSOH model or something like that.

**Comment by valentinsulzer** [comment]
It's probably something in pybamm, my guess is the `set_initial_conditions_from` function, but hard to tell exactly what
